### PR TITLE
Implements System-Token Validation

### PR DESCRIPTION
**Pull request by lcaparroz**
## Description

This pull request implements the server-side handling of the `System-Token` header on requests by clients that support the feature.

## Change Type

*Please select the correct option.*

- [ ] **Bug Fix** (a non-breaking change which fixes an issue)
- [x] **New Feature** (a non-breaking change which adds new functionality)
- [ ] **Documentation Update** (a change which only updates documentation)

## Checklist

*Please check off each item if the requirement is met.*

- [x] I have verified that my code follows RMT's coding standards with `rubocop`.
- [ x I have reviewed my own code and believe that it's ready for an external review.
- [x] I have provided comments for any hard-to-understand code.
- [ ] I have documented the `MANUAL.md` file with any changes to the user experience.
- [x] RMT's test coverage remains at 100%.
- [ ] If my changes are non-trivial, I have added a changelog entry to notify users at `package/obs/rmt-server.changes`.

### Setup

1. In your terminal shell, change directory to your local RMT clone repository;
2. Start a new Rails console with `bin/rails console`;
3. Create a new system in your local RMT instance:

   ```ruby
   System.create!(login: 'system-token-test', password: 'password')
   ```

4. Exit the Rails console, and start a Rails server with:

   ```sh
   bin/rails server -b localhost -p 4224
   ```

5. Start a new shell session, where you can execute the `curl` command.

### Scenarios

All the following scenarios and

#### Case 1: Invalid Credentials

A system tries to authenticate with the API, but no systems with the given
credentials have been found.

##### Execution

1. In your shell, execute:

   ```sh
   curl -v 'http://0.0.0.0:4224/connect/systems/products' \
     -H 'accept: application/json' \
     -u 'system-token-test:wrong-password'
   ```

##### Expectation

1. The Rails server log output should contain messages like:

   ```
   I, [2022-07-14T21:10:01.179508 #1433]  INFO -- : Could not find system with login "system-token-test" and password "passworda"
   ```

2. Check the `curl` command output should contain messages like:

   ```sh
   < HTTP/1.1 401 Unauthorized
   ```

#### Case 2: Successful Authentication without `System-Token` Header

A system tries to authenticate with the API, without sending the `System-Token`
header, but with credentials that match *at least* on system. This is a typical
scenario for older SUSEConnect clients versions.

##### Execution

1. In your shell, execute:

   ```sh
   curl -v 'http://0.0.0.0:4224/connect/systems/products' \
     -H 'accept: application/json' \
     -u 'system-token-test:password'
   ```

##### Expectation

1. The Rails server log output should contain messages like:

   ```
   I, [2022-07-14T21:21:15.751777 #1427]  INFO -- : System with login "system-token-test" authenticated without token header
   ```

2. Check the `curl` command output should contain messages like:

   ```sh
   < HTTP/1.1 422 Unprocessable Entity
   ```

The system successfully authenticated , but there are no products associated to
it yet, hence the HTTP 422 response.

#### Case 3: Successful Authentication with Matching Empty `System-Token` Header

A system tries to authenticate with the API, sending the `System-Token` header
with an empty value, and there is only one system with the given credentials and
an empty token. This is a typical scenario for systems that have just announced
themselves to an RMT instance (first time they call the API).

##### Execution

1. In your shell, execute:

   ```sh
   curl -v 'http://0.0.0.0:4224/connect/systems/products' \
     -H 'System-Token;' \
     -H 'accept: application/json' \
     -u 'system-token-test:password'
   ```

##### Expectation

1. The Rails server log output should contain messages like:

   ```
   I, [2022-07-14T21:34:31.636182 #1427]  INFO -- : System with login "system-token-test" authenticated with token ""
   ```

2. Check the `curl` command output should contain messages like:

   ```sh
   < HTTP/1.1 422 Unprocessable Entity
   < System-Token: fd9f911b-2475-4bb5-ac8f-a49b78a269fa
   ```

The system successfully authenticated, and a new system token has been generated
at server-side and returned to the client. In the example above, it would be
`fd9f911b-2475-4bb5-ac8f-a49b78a269fa`

##### Post-Execution

1. Take note of the returned `System-Token` header so it can be used in the next
   case/scenario.

#### Case 4: Successful Authentication with Matching `System-Token` Header

A system tries to authenticate with the API, sending a `System-Token` with a
value which, combined with the credentials, uniquely matches a system in the
database.

##### Execution

1. In your shell, execute:

   ```sh
   curl -v 'http://0.0.0.0:4224/connect/systems/products' \
     -H 'System-Token: fd9f911b-2475-4bb5-ac8f-a49b78a269fa' \
     -H 'accept: application/json' \
     -u 'system-token-test:password'
   ```

##### Expectation

1. The Rails server log output should contain messages like:

   ```
   I, [2022-07-14T21:43:46.741382 #1433]  INFO -- : System with login "system-token-test" authenticated with token "fd9f911b-2475-4bb5-ac8f-a49b78a269fa"
   ```

2. Check the `curl` command output should contain messages like:

   ```sh
   < HTTP/1.1 422 Unprocessable Entity
   < System-Token: 5899d5f8-61ea-491a-accd-593ba7054927
   ```

The system successfully authenticated with a valid system token, so a new one is
generated server-side and returned to the client.

#### Case 5: Successful Authentication with Mismatching Empty `System-Token` Header

A system tries to authenticate with the API, using credentials that matches at
least one system, but with a `System-Token` header value which is not valid.
This is a typical scenario for a cloned system.

##### Execution

1. In your shell, execute:

   ```sh
   curl -v 'http://0.0.0.0:4224/connect/systems/products' \
     -H 'System-Token: invalid-system-token' \
     -H 'accept: application/json' \
     -u 'system-token-test:password'
   ```

##### Expectation

1. The Rails server log output should contain messages like:

   ```
   I, [2022-07-14T21:50:23.217769 #1427]  INFO -- : System with login "system-token-test" (ID 2025683) authenticated and duplicated from ID 2025682 due to token mismatch
   ```

2. Check the `curl` command output should contain messages like:

   ```sh
   < HTTP/1.1 422 Unprocessable Entity
   < System-Token: 93d0c1ad-fb1b-49e4-b764-ce21fcbb4fae
   ```

The system successfully authenticated, but it was duplicated from another
system, so a new system token is generated server-side and returned to the
client. Notice the client doesn't receive any information about the duplication.

#### Case 6: Successful Authentication with Mismatching `System-Token` Header

A system tries to authenticate with the API using credentials that **only**
matches systems with full token, but it sends an empty `System-Token` header.
This is a typical scenario for a cloned system.

##### Execution

1. In your shell, execute:

   ```sh
   curl -v 'http://0.0.0.0:4224/connect/systems/products' \
     -H 'System-Token;' \
     -H 'accept: application/json' \
     -u 'system-token-test:password'
   ```

##### Expectation

1. The Rails server log output should contain messages like:

   ```
   I, [2022-07-14T21:55:15.191731 #1427]  INFO -- : System with login "system-token-test" (ID 2025684) authenticated and duplicated from ID 2025682 due to token mismatch
   ```

2. Check the `curl` command output should contain messages like:

   ```sh
   < HTTP/1.1 422 Unprocessable Entity
   < System-Token: 64c65406-87a6-44cc-8868-4ecfe9e52fef
   ```

The system successfully authenticated, but it was duplicated from another
system, so a new system token is generated server-side and returned to the
client. Notice the client doesn't receive any information about the duplication.

### CODE REVIEW

**Comment by lcaparroz** [comment]
[[URAS] Add system token handling to RMT](https://trello.com/c/sB5sI7SC/2441-uras-add-system-token-handling-to-rmt)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/261528/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 8b1794885ba76229a5ecc6e5628a224b790c466d into 454855918e87bc5de7d526e497708e2e57e75488)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/261802/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 722ee9d7f7b06772a10e05698093e34b37fe863c into 454855918e87bc5de7d526e497708e2e57e75488)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/261803/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 92660a94e1fb93a6721125bc0165d6b5701b631e into 454855918e87bc5de7d526e497708e2e57e75488)

**Review by ngetahun — CHANGES_REQUESTED** [review]

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/262846/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 92660a94e1fb93a6721125bc0165d6b5701b631e into 71eecbdb59eda8f6be8d28244b7997e1bb6412c9)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/262987/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 04b5f713a3910fd1a419f7fda1a0641e6c91be62 into 71eecbdb59eda8f6be8d28244b7997e1bb6412c9)

File: spec/support/shared_examples/api_system_token.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
+shared_examples 'responds with a new token' do
```
[the remark this answers is not in the document]
Comment (lcaparroz) [reply spec/support/shared_examples/api_system_token.rb]: Done at [d68a4bb](https://github.com/SUSE/rmt/pull/913/commits/d68a4bb8262ea474ad019039bc5b1252a89fa146).

**Review by ngetahun — APPROVED** [review]
LGTM!

**Pull request merged**